### ISSUE TITLE
Favorited avatar not always loading automatically

### ISSUE THREAD

**Issue by gentlecolts** [Yellow-Dog-Man/Resonite-Issues#123]
### Describe the bug?

Often I'll find myself in the default head/hands avatar, despite having the same one favorited that's been set for ages.  Usually it loads correctly, no idea what causes it to not load

### To Reproduce

Haven't been able to consistently do so

### Expected behavior

Favorited avatar should always load when not spectator

### Resonite Version Number

2023.10.13.202

### What Platforms does this occur on?

Windows, Linux

### Reporters

zangooseoo

**Comment by shadowpanther** [comment]
The favourited avatar usually does not load if the session's host is away from the session. It loads when the host comes back.

Some worlds also have roles set up such that you can't use custom avatars.

**Comment by gentlecolts** [comment]
I'm aware of that particular bug, and perhaps it should be covered under this, but I'm occasionally seeing it in cases where that shouldn't happen, such as on headless servers

**Comment by Nammi-namm** [comment]
> I'm aware of that particular bug, and perhaps it should be covered under this, but I'm occasionally seeing it in cases where that shouldn't happen, such as on headless servers

Neither is a bug. It's intended behaviour. 

It's intentional that people joining don't have their avatar load when the host is away in a different session. They will load once the host comes back.

Regarding users joining not having their favourited cloud avatar spawn with them in a world where the host is present or is being hosted by a headless. This is intended behaviour. In root under Controllers, there is a checkbox for LoadCloudAvatars: 
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/3219044/6ffe1bcc-a0ea-4849-bbd2-6c4d9df9a2a0)
This is not role specific, it applies to all users joining the session. Many headlesses have this disabled for joining users to avoid lag. Sometimes they tell you to choose an optimised avatar in the entrance and have that disabled so it doesn't load anyone's favourite unoptimised avatar.

I would confirm with the host of any headless you've seen this happen on to confirm its actually a bug and not an intentional setting.

**Comment by gentlecolts** [comment]
I am aware of these cases, however my avatar is occasionally failing to load in cases unrelated to these expected scenarios.  Just now, on a non-headless instance, with a user host who was present, who has done nothing in the world to prevent avatar loading in the world, I loaded in the default avatar.  The case I was specifically mentioning earlier thatwas on a headless was in a world in which I know did not have that set as well (I made the world)

**Comment by shadowpanther** [comment]
> It's intentional that people joining don't have their avatar load when the host is away in a different session. They will load once the host comes back.

I don't think this behaviour is _intentional_, but that's how sessions currently work when the host is away. A separate issue should be mad about this.

The point of this issue, though, is the case when neither of these two causes apply. I've just mentioned those because it was not clear from the original description.

**Comment by shiftyscales** [comment]
Have you been able to determine reliable ways to reproduce this issue, @gentlecolts? E.g. have you had it occur on either self-hosted or user-hosted sessions in which the host is definitely present / focused on that world?

**Comment by gentlecolts** [comment]
> Have you been able to determine reliable ways to reproduce this issue, [@gentlecolts](https://github.com/gentlecolts)

If I'll be completely honest, I haven't been specifically looking for it.  I've kind of just come to expect I'll sometimes just not be in my avatar when I load a session, with no discernible rhyme or reason.  The ability for hosts/map makers to control whether or not this happens muddies this issue, as I can never just know whether I start a session in the default avatar because the host said so, because the host was afk, or because of some freak who knows what

**Comment by shiftyscales** [comment]
I see- it would help a lot if we can be certain that there is a particular replication case for this issue that isn't related to all of the known / existing mechanisms / reasons for why one's avatar might not load when entering a session.

If you happen to catch this occurring again and can determine it is definitely not related to any known reason why this would occur- please mark the moment with the F3 key to timestamp it, and include log files- ideally from both the host you are joining and your client if possible.

**Comment by gentlecolts** [comment]
> please mark the moment with the F3 key to timestamp it

I wasn't aware of this feature, thanks for the tip, I'll keep it in mind!

**Comment by Lixquid** [comment]
Apologies on just jumping in, but I believe @RyuviTheViali explained to me once what may be happening here (and they can jump in if I'm wrong); when you first join a world, the host gets a short (I think tens of seconds?) window to automatically spawn and apply your default set avatar (since I believe the host doesn't normally get free license to just rummage around in your cloud inventory), and if you take longer than that to load in or deal with other network matters, the license expires and you're left with the default avatar.

I typically have this happen sometimes when loading into a world with a lot of existing data or with a large number of people in it; does this match the times you were experiencing this behaviour?

**Comment by ko-tengu** [comment]
I didn't actually realize we had a generic version of this since there's a more detailed one that has the problem, as mentioned, #1386 . I find it happens most often when connecting to a server with a higher ping, every time I connect to one of my European friends I get this, about 130ms average ping to them from central US.

**Comment by stiefeljackal** [comment]
> Apologies on just jumping in, but I believe [@RyuviTheViali](https://github.com/RyuviTheViali) explained to me once what may be happening here (and they can jump in if I'm wrong); when you first join a world, the host gets a short (I think tens of seconds?) window to automatically spawn and apply your default set avatar (since I believe the host doesn't normally get free license to just rummage around in your cloud inventory), and if you take longer than that to load in or deal with other network matters, the license expires and you're left with the default avatar.

This is roughly correct. When it is time for the avatar to spawn via `CommonAvatarBuilder`, there is a 30 second window for the spawned user to grant an access key for the host to use to access and to spawn the avatar. If the access key is not given to the host once 30 seconds have passed, the builder will no longer wait and spawn the user in the world with the default avatar equipped.

**Comment by shiftyscales** [comment]
Hmm- if we are fairly certain that #1386 is the likely cause here- I think I will close this issue in favour of that one. If there do turn out to be any other known causes that aren't addressed, please feel free to open a new issue detailing other conditions under which this issue occurs.

**Comment by ko-tengu** [comment]
It seems that this bug may not entirely be related to the other issue, because it is still happening after the update that should have fixed the avatar timeout error. I don't have the advanced logging functions, but I have noticed that I will fail to load into my favorited avatar nearly 100% of the time when I have a higher ping to the session host. I have a friend that I have a consistent 100ms ping to and I will never enter their sessions in the correct avatar.

I marked the log with the F3 debug call, since this is a partial log, if needed I can upload the full log when I log off for the day.

[BUFU - 2024.10.8.1349 - 2024-10-17 18_07_40.log](https://github.com/user-attachments/files/17423813/BUFU.-.2024.10.8.1349.-.2024-10-17.18_07_40.log)

**Comment by shiftyscales** [comment]
The AvatarAccessKey is now set to last five minutes- had you waited at least that long, @ko-tengu? And there is no chance it is any other possible explanations / known causes? E.g. the host having the world unfocused, or the session disallowing cloud avatars?

**Comment by ko-tengu** [comment]
Waiting about 6 minutes in this session tonight and my avatar never spawned in until I manually spawned and switched to it. Log marked with f3 at the about six minute mark.
[BUFU - 2024.10.8.1349 - 2024-10-19 22_03_34.log](https://github.com/user-attachments/files/17448130/BUFU.-.2024.10.8.1349.-.2024-10-19.22_03_34.log)

**Comment by shiftyscales** [comment]
Seeking input from @Frooxius - following up from #1386, it seems there may be some additional conditions under which favourited avatars are not loading correctly- are there any additional diagnostics that could be added which could help possibly identify other points of failure?

There are known causes such as #187, or world which purposefully disable cloud spawned avatars - but based on @ko-tengu's feedback above- it sounds like there may be other circumstances in which avatars will fail to load properly sometimes?

**Comment by Frooxius** [comment]
@ko-tengu @shiftyscales We will need logs from the host for this, because they are the one who actually spawns the avatar in this case.
